Mathlib/Algebra/Ring/SumsOfSquares.lean

```diff
@@ -0,0 +1,97 @@
+/-
+Copyright (c) 2024 Florent Schaffhauser. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Florent Schaffhauser
+-/
+
+import Mathlib.Algebra.Ring.Defs
+import Mathlib.Algebra.Order.Ring.Defs
+
+/-!
+# Sums of squares
+
+We introduce the type sums of squares in a semiring `R` as the subtype of `R` whose terms are
+sums of squares in `R`. The associated predicate `isSumSq : R → Prop` is defined inductively:
+`0 : R` is a sum of squares and if `S` is a sum of squares, then for all `a : R`, `a ^ 2 + S` is a
+sum of squares in `R`.
+
+## Main declaration
+
+- The type `SumSqIn R`: it is the type of sums of squares in `R`, where `R` is a semiring.
+
+## Auxiliary declarations
+
+- The predicate `isSumSq : R → Prop`: the predicate used to define the type `SumSqIn R` as a subtype
+of `R`.
+
+## Theorems
+
+- `isSumSq.add`: if `S₁` and `S₂` are sums of squares in a semiring `R`, then so is `S₁ + S₂`.
+- `SumSq.nonneg`: sums of squares are non-negative.
+
+## Instances
+
+- An `Add` instance on the type of sums of squares in a semiring `R`.
+
+-/
+
+/--
+In a semiring `R`, the property of being a sum of squares is an inductive predicate: `0 : R` is a
+sum of squares and if `S` is a sum of squares, then for all `a : R`, `a ^ 2 + S` is a sum of squares
+in `R`.
+-/
+
+@[mk_iff]
+inductive isSumSq {R : Type*} [Add R] [Mul R] [Zero R] : R → Prop
+  | zero                              : isSumSq 0
+  | sq_add (x S : R) (pS : isSumSq S) : isSumSq (x * x + S)
+
+/--
+The type of sums of squares in a semiring `R` is the subtype of `R` defined by the predicate
+`isSumSq : R → Prop`. It  can be defined as a structure whose terms are dependent pairs `⟨x, hx⟩`
+where `x : R` and `hx` is a proof of  the proposition `isSumSq x`.
+-/
+
+structure SumSqIn (R : Type*) [Add R] [Mul R] [Zero R] where
+  /-- `val` is a term in `R` and `ppty` is a proof of the proposition `isSumSq val` -/
+  val  : R
+  ppty : isSumSq val
+
+/--
+If `S1` and `S2` are sums of squares in a semiring `R`, then `S1 + S2` is a sum of squares in `R`.
+-/
+
+/-
+*TODO*: `isSumSq.add` could be rewritten with weaker assumptions on `R` (we only need to
```

Comment:
There should be a way to generalise `SumSq.add` from semirings to types `R` with `AddZeroClass R` and `AddSemigroup R` instances (for some reason, having just these two does not work, there must be a distributivity property, or some other compatibility condition between `Add R` and `Mul R` used somewhere in the file and I am not sure what is the minimal framework for this; certainly in semirings it works).

Reply:
Let me see.

Reply:
```lean
theorem isSumSq.add {R : Type*} [Mul R] [AddMonoid R] {S1 S2 : R} (h1 : isSumSq S1) (h2 : isSumSq S2) :
    isSumSq (S1 + S2) := by
  induction h1 with
  | zero             => rwa [zero_add S2]
  | sq_add a S hS ih =>
    rw [add_assoc]
    exact sq_add _ _ ih
```

Reply:
`AddZeroClass` plus `AddSemigroup` is exactly `AddMonoid` (not by definition, but it seems trivially true to me)

Reply:
Yes! If I use `[AddMonoid R] [Mul R]`, it works 🙏🏻 

**EDIT:** Oh, I didn't see your code above 😓 Thanks!